lib/evmone_precompiles/kzg.cpp

```diff
@@ -0,0 +1,155 @@
+std::optional<blst_p1_affine> validate_G1(std::span<const std::byte, 48> b) noexcept
+{
+    blst_p1_affine r;
+    if (blst_p1_uncompress(&r, reinterpret_cast<const uint8_t*>(b.data())) != BLST_SUCCESS)
+        return std::nullopt;
+    if (!blst_p1_affine_in_g1(&r))  // Subgroup check is required by the spec but not testable.
+        return std::nullopt;
+    return r;
+}
+
+blst_p1_affine add_or_double(const blst_p1_affine& p, const blst_p1& q) noexcept
+{
+    blst_p1 r;
+    blst_p1_add_or_double_affine(&r, &q, &p);
+    blst_p1_affine ra;
+    blst_p1_to_affine(&ra, &r);
+    return ra;
+}
+
+blst_p1 mult(const blst_p1& p, const blst_scalar& v) noexcept
+{
+    blst_p1 r;
+    blst_p1_mult(&r, &p, v.b, 255);
+    return r;
+}
+
+blst_p2_affine add(const blst_p2_affine& p, const blst_p2& q) noexcept
+{
+    blst_p2 r;
+    blst_p2_add_affine(&r, &q, &p);
+    blst_p2_affine ra;
+    blst_p2_to_affine(&ra, &r);
+    return ra;
+}
+
+blst_p2 mult(const blst_p2& p, const blst_scalar& v) noexcept
+{
+    blst_p2 r;
+    blst_p2_mult(&r, &p, v.b, 255);
+    return r;
+}
+
+bool pairings_verify(
+    const blst_p1_affine& a1, const blst_p1_affine& b1, const blst_p2_affine& b2) noexcept
+{
+    blst_fp12 left;
+    blst_aggregated_in_g1(&left, &a1);
+    blst_fp12 right;
+    blst_miller_loop(&right, &b2, &b1);
+    return blst_fp12_finalverify(&left, &right);
+}
+}  // namespace
+
+bool kzg_verify_proof(const std::byte versioned_hash[32], const std::byte z[32],
+    const std::byte y[32], const std::byte commitment[48], const std::byte proof[48]) noexcept
+{
+    std::byte computed_versioned_hash[32];
+    sha256(computed_versioned_hash, commitment, 48);
+    computed_versioned_hash[0] = std::byte{0x01};
+    if (!std::ranges::equal(std::span{versioned_hash, 32}, computed_versioned_hash))
+        return false;
+
+    // Load and validate scalars z and y.
+    // TODO(C++26): The span construction can be done as std::snap(z, std::c_<32>).
+    const auto zz = validate_scalar(std::span<const std::byte, 32>{z, 32});
+    if (!zz)
+        return false;
+    const auto yy = validate_scalar(std::span<const std::byte, 32>{y, 32});
+    if (!yy)
+        return false;
+
+    // Uncompress and validate the points C (representing the polynomial commitment)
+    // and Pi (representing the proof). They both are valid to be points at infinity
+    // when they prove a commitment to a constant polynomial,
+    // see https://hackmd.io/@kevaundray/kzg-is-zero-proof-sound
+    const auto C = validate_G1(std::span<const std::byte, 48>{commitment, 48});
+    if (!C)
+        return false;
+    const auto Pi = validate_G1(std::span<const std::byte, 48>{proof, 48});
+    if (!Pi)
+        return false;
+
+    // Compute -Y as [y * -1]₁.
+    const auto neg_Y = mult(G1_GENERATOR_NEGATIVE, *yy);
+
+    // Compute C - Y. It can happen that C == -Y so doubling may be needed.
+    const auto C_sub_Y = add_or_double(*C, neg_Y);
```

Comment:
I would probably always use add_or_double everywhere as I'm guessing the performance impact is negligible

Reply:
Done.